File: ai/src/wiki_mcp/tools/scope.py
"""Replaces lucas-llmwiki `mcp/tools/list.py`.

Upstream let the agent create and rename knowledge bases. Here it cannot: a scope
is a visibility partition derived from departments (DR-018, `wiki_scope`), and
only the backend creates one. What is left is enumeration, so the agent learns
which `scope` value to pass.
"""

from mcp.server.fastmcp import Context, FastMCP


def register(mcp: FastMCP, get_scope_key, fs_factory) -> None:

    @mcp.tool(
        name="list_scopes",
        description=(
            "이 서버가 다루는 공개 범위를 나열한다.\n\n"
            "다른 모든 도구가 `scope` 인자를 받는다 — 유효한 값을 모를 때 이걸 먼저 부른다. "
            "범위는 부서 조합에서 나온 공개 단위이고 에이전트가 만들 수 없다."
        ),
    )
    async def list_scopes(ctx: Context) -> str:
        fs = fs_factory(get_scope_key(ctx))
        scopes = await fs.list_scopes()
        if not scopes:
            return "범위가 없다. 백엔드가 먼저 범위를 만들어야 한다."
        return "\n".join(
            f"- **{s['scope_key']}** "
            f"({'전사' if s['visibility_type'] == 'ALL' else '부서 한정'}) — "
            f"원본문서 {s.get('source_count', 0)}건, 위키 {s.get('page_count', 0)}페이지, "
            f"목차 `{s['index_path']}`"
            for s in scopes
        )

File: frontend/src/features/document/scope.js
// 공개 범위 선택 → scopeKey 문자열.
// 서버가 최종 scopeKey를 계산하지만(업로드 응답으로 되돌려줌), 카테고리 목록을
// scopeKey별로 조회해야 해서 클라이언트도 같은 규칙으로 파생값을 만든다.
// 규칙(FR-DOC-002): 부서 ID 중복 제거 후 오름차순 정렬, "D1-D2" 형태. 전체 공개는 'ALL'.
export function buildScopeKey(visibilityType, departmentIds = []) {
  if (visibilityType === 'all') return 'ALL'
  const ids = [...new Set(departmentIds.map(String))].sort((a, b) => Number(a) - Number(b))
  return ids.length ? ids.map((id) => `D${id}`).join('-') : ''
}
